### ISSUE TITLE
Possibility to assign Devices Groups to a DevOps Pipeline

### ISSUE THREAD

**Issue by MarianRaphael** [FlowFuse/flowfuse#2989]
## User Story

As a FlowFuse user,
I want the ability to push a Snapshot through a DevOps Pipeline to a groups of devices,
So that I can efficiently manage and update multiple devices simultaneously 

```[tasklist]
### Requirements
- [ ] https://github.com/FlowFuse/flowfuse/issues/2902
- [ ] https://github.com/FlowFuse/flowfuse/issues/2756
- [ ] https://github.com/FlowFuse/flowfuse/issues/3183
```

**Comment by MarianRaphael** [comment]
Next iteration: https://github.com/FlowFuse/flowfuse/issues/3142

**Comment by Steve-Mcl** [comment]
@MarianRaphael, @joepavitt , @Yndira-FlowForge 


Questions on design specifically around "Device Groups" that will soon block me - looking for direction please?

![Image](https://github.com/FlowFuse/flowfuse/assets/44235289/37097823-8e0b-4b73-a3e1-4758e4cb9b72)

1. Device Group colour?
2. What will we display in the stage?
    * Considering there could be a large about of devices, perhaps simple stats like:
       *  `Device Count` pill 
       * `Targeted Count` pill (the amount of devices with target snapshot set to that of the last pipeline deploy)
       * `Active Count` (the number of devices in the group that have the Active snapshot set to that of the last pipeline deploy)
3. Should we have a stage AFTER a Device Group stage?  Does it make sense?
4. If yes to above ^ then what are the deployment actions and can this me a later iteration or is it MVP?

**Comment by joepavitt** [comment]
> 1. Device Group colour?

I had assumed the same at the "Device Icon". The blue we use as a link, purple is for the dev mode. Green/Yellow have traffic light connotations, so I'd say keep the same. Happy to be overruled by @Yndira-FlowForge though.

> 2. What will we display in the stage?

I'd agree with your ideas here. Although a user is then inevitably going to want to see which devices haven't been deployed to yet, although I guess that's accessible by going to the Device Group page anyway?

> 3. Should we have a stage AFTER a Device Group stage? Does it make sense?

Good question as it's not clear what the selected action could do. There are no Device Group scoped Snapshots or anything, so I'd say, for now, nothing can go after a Device Group in a Pipeline. Thoughts @MarianRaphael?

**Comment by MarianRaphael** [comment]
> Should we have a stage AFTER a Device Group stage? Does it make sense?

To make it as simple as possible in this iteration, no

> Although a user is then inevitably going to want to see which devices haven't been deployed to yet

I have a follow-up story for this need: https://github.com/FlowFuse/flowfuse/issues/3142
But I agree just a small notification with the different numbers is a good idea: 

- Device Count pill
- Targeted Count pill 
- Active Count

**Comment by Steve-Mcl** [comment]
> > 1. Device Group colour?
> 
> I had assumed the same at the "Device Icon". The blue we use as a link, purple is for the dev mode. Green/Yellow have traffic light connotations, so I'd say keep the same. Happy to be overruled by @Yndira-FlowForge though.

@Yndira-FlowForge can you comment please?

IMHO, the device and device-group icons are so similar that habing the same colour makes it a little difficult to distinguish 

![image](https://github.com/FlowFuse/flowfuse/assets/44235289/3a61dea5-a905-4955-be50-fe4c6ef89bfa)

**Comment by Steve-Mcl** [comment]
RE: Device Group status info

This is what I ended up achieving:

### V1: Successful deployment
![image](https://github.com/FlowFuse/flowfuse/assets/44235289/3c86ede5-ebeb-4206-a9f7-f3c00153b2ff)


### V1: Added another device to Device Group

![image](https://github.com/FlowFuse/flowfuse/assets/44235289/b3bf53dd-8683-488a-a653-a4e9f623d08d)


### V1: Partial deployment
![image](https://github.com/FlowFuse/flowfuse/assets/44235289/a193f2dc-86df-41d2-86c8-2993d1c9f958)


<br>

This was discussed with Joe. 

Although Joe knew what the numbers were, his instinct was to "add them up" and felt it was distracting / unclear.

After some  discussion, Joe has asked that the "Target" and "Developer Mode" counts be removed.  

This leaves us with this:


### V2: A full pipeline deploy 

![image](https://github.com/FlowFuse/flowfuse/assets/44235289/519f2ee4-1b1f-4252-b290-bfdd195870b0)

### V2: A NON fully active pipeline deployed group (status by inference) 

![image](https://github.com/FlowFuse/flowfuse/assets/44235289/d76ff173-7cc7-46f8-bed0-3ff19a24ddae)



---

@joepavitt tagged for his thoughts (as discussed)

@Yndira-FlowForge for your thoughts alternative design direction

**Comment by joepavitt** [comment]
We should also have a tooltip on the "3" to detail this a little further.

The listing of the states beyond just "Active" is confusing as it still doesn't give the full picture. My reasoning for just showing Active is that's what the user cares about here, so put it front and centre. The others were distracting.

To investigate why, they would need to click on the Device Group, and then in the "Devices" view, we need to list target/Active snapshot as per the existing Devices Browser. I know this is a followup action though, and not part of 1st iteration here

**Comment by Steve-Mcl** [comment]
> We should also have a tooltip on the "3" to detail this a little further.

![image](https://github.com/FlowFuse/flowfuse/assets/44235289/8408b9dc-f105-475a-b3de-e1353eacd21c)


If the device count matches the active count, there will be no tootip

_open to suggestions on wording_

**Comment by joepavitt** [comment]
> there will be no tootip

"All Devices running latest snapshot" in this case would be clearer.

> open to suggestions on wording

Think as you have it is fine for now, but we change it to _"Some devices not running latest Snapshot. View Device Group for more information"_ or something along those lines - maybe too wordy.

**Comment by Steve-Mcl** [comment]
> or something along those lines - maybe too wordy

After a nights sleep and going back n forth, I have stumbled on the following wording that _feels_ more "related" to piplines (since the status badge is driven by the pipeline deployed targetSnapshotId) 🤞 


![image](https://github.com/FlowFuse/flowfuse/assets/44235289/50c04e54-26e6-42ac-8483-dd107de02224)


![image](https://github.com/FlowFuse/flowfuse/assets/44235289/39de8690-757b-4d19-aac1-cc4a148fedc9)

**Comment by joepavitt** [comment]
nailed it @Steve-Mcl 👍

**Comment by Steve-Mcl** [comment]
Staging verification: 20/12/2023


## "Steve on Staging" team


### Pipeline ops
- [x] Cannot Add 1st stage in pipeline with a device group
- [x] Can Add 2nd stage with a device group
   - [x] Team Audit Log entry present
- [x] Cannot add any type of stage _after_ a Device Group (no add stage button)
- [x] Can change a last stage instance/device stage to a device group stage
- [x] Can NOT change a middle stage (e.g. stage 2 in a 3 stage pipeline) a device group stage
   - [x] Expect toast error "Request Failed: A Device Group can only set on the last stage"
- [x] Can deploy to a device group
  - [x] All devices get immediate update
  - [x] Team Audit Log entry present


### Issues noted (follow up issues/tasks to be raised for each)
1. Instance info missing if suspended: PR - https://github.com/FlowFuse/flowfuse/pull/3252
1. Empty State is displayed on pipeline when first navigated to
1. Disabling Device Groups on team type leaves configured device group in a pipeline stage (still works, displays correctly etc, needs design discussion)
